libs/core/gasnet_base/src/gasnet_environment.cpp

```diff
@@ -0,0 +1,701 @@
+    hpx::mutex gasnet_environment::dshm_mut{};
+    hpx::mutex gasnet_environment::mtx_{};
+    bool gasnet_environment::enabled_ = false;
+    bool gasnet_environment::has_called_init_ = false;
+    int gasnet_environment::provided_threading_flag_ = GASNET_PAR;
+    int gasnet_environment::is_initialized_ = -1;
+    int gasnet_environment::init_val_ = GASNET_ERR_RESOURCE;
+    hpx::mutex* gasnet_environment::segment_mutex = nullptr;
+    gasnet_seginfo_t* gasnet_environment::segments = nullptr;
+
+    ///////////////////////////////////////////////////////////////////////////
+    int gasnet_environment::init(int* argc, char*** argv, const int minimal,
+        const int required, int& provided)
+    {
+        if (!has_called_init_)
+        {
+            gasnet_environment::init_val_ = gasnet_init(argc, argv);
+            has_called_init_ = true;
+        }
+
+        if (gasnet_environment::init_val_ == GASNET_ERR_NOT_INIT)
+        {
+            HPX_THROW_EXCEPTION(invalid_status,
+                "hpx::util::gasnet_environment::init",
+                "GASNET initialization error");
+        }
+        else if (gasnet_environment::init_val_ == GASNET_ERR_RESOURCE)
+        {
+            HPX_THROW_EXCEPTION(invalid_status,
+                "hpx::util::gasnet_environment::init", "GASNET resource error");
+        }
+        else if (gasnet_environment::init_val_ == GASNET_ERR_BAD_ARG)
+        {
+            HPX_THROW_EXCEPTION(invalid_status,
+                "hpx::util::gasnet_environment::init",
+                "GASNET bad argument error");
+        }
+        else if (gasnet_environment::init_val_ == GASNET_ERR_NOT_READY)
+        {
+            HPX_THROW_EXCEPTION(invalid_status,
+                "hpx::util::gasnet_environment::init",
+                "GASNET not ready error");
+        }
+
+        if (provided < minimal)
+        {
+            HPX_THROW_EXCEPTION(invalid_status,
+                "hpx::util::gasnet_environment::init",
+                "GASNET doesn't provide minimal requested thread level");
+        }
+
+        if (gasnet_attach(nullptr, 0, gasnet_getMaxLocalSegmentSize(), 0) !=
+            GASNET_OK)
+        {
+            HPX_THROW_EXCEPTION(invalid_status,
+                "hpx::util::gasnet_environment::init",
+                "GASNET failed to attach to memory");
+        }
+
+        // create a number of segments equal to the number of hardware
+        // threads per machine (locality)
+        //
+        gasnet_environment::segments = new gasnet_seginfo_t[size()];
+        gasnet_environment::segment_mutex = new hpx::mutex[size()];
+
+        GASNET_Safe(gasnet_getSegmentInfo(segments, size()));
+
+        int retval;
+
+        gasnet_barrier_notify(0, GASNET_BARRIERFLAG_ANONYMOUS);
+        while ((retval = gasnet_barrier_try(0, GASNET_BARRIERFLAG_ANONYMOUS)) ==
+            GASNET_ERR_NOT_READY)
+        {
+        }
+        gasnet_barrier_try(0, GASNET_BARRIERFLAG_ANONYMOUS);
```

Comment:
This is one of a large number of user errors that a GASNet library built in debug mode (`configure --enable-debug`) automatically detects and reports as a fatal error. The fact this code is here tells me this PR has never been tested against GASNet debug mode (a low-performance mode whose life purpose is to detect these types of errors). 

Please perform some correctness testing against a GASNet library built in debug mode to shake out other latent defects like this one. 

Documentation from [GASNet README](https://gasnet.lbl.gov/dist-ex/README):
>  --enable-debug - 
> build GASNet in a debugging mode. This turns on C-level       debugger options and also enables extensive error and sanity checking       system-wide, which is highly recommended for developing and debugging       GASNet clients (but should NEVER be used for performance testing).       --enable-debug also implies --enable-{trace,stats,debug-malloc},       but these can still be selectively --disable'd.

Reply:
Sounds good. Will make updates.